types/index.d.ts

```diff
@@ -0,0 +1,130 @@
+interface PortBuilderType {
+    type: string;
+    name?: string;
+    label?: string;
+    noControls?: boolean;
+    color?: string;
+    hidePort?: boolean;
+    controls?: boolean;
+}
+
+export interface NodeType {
+    type: string;
+    label?: string;
+    initialWidth?: number;
+    inputs?: ((ports: PortTypes) => PortBuilderType[]) | PortBuilderType[];
+    outputs?: ((ports: PortTypes) => PortBuilderType[]) | PortBuilderType[];
+    root?: boolean;
+    addable?: boolean;
+    deletable?: boolean;
+    description?: string;
+    sortIndex?: number;
+}
+
+interface PortProps {
+    label: string;
+    inputLabel: string;
+    name: string;
+    portName: string;
+    defaultValue: any;
+    inputData: any;
+}
+
+interface Control {
+    type: string;
+    name?: string;
+    label?: string;
+    defaultValue?: any | any[];
+    setValue?: (oldData: any, newData: any) => any;
+    options?: any[];
+    getOptions?: (any) => any;
+    placeholder?: any;
+    render?: (
+        data: any,
+        onChange: (data: any) => void,
+        context: any,
+        redraw: () => void,
+        portProps: PortProps,
+        inputData: any
+      ) => JSX.Element;
+}
+
+export interface PortType {
+    type: string;
+    name: string;
+    label?: string;
+    acceptTypes?: string[];
+    color: string;
+    hidePort?: boolean;
+    controls?: Control[]
+}
+
+interface NodeBase {
+    id: string;
+    x: number;
+    y: number;
+    width: number;
+    height: number;
+    color: string;
+}
+
+interface Connection {
+    nodeId: string;
+    portName: string;
+}
+
+export interface Node extends NodeBase {
+    type: string;
+    connections: {
+        inputs: {[id: string]: Connection};
+        outputs: {[id: string]: Connection};
+    };
+    inputData: {[id: string]: {string?: string, boolean?: boolean, number?: number}};
+}
+
+export interface Comment extends NodeBase {
+    text: string;
+}
+
+export interface DefaultNode {
+    type: string;
+    x: number;
+    y: number;
+}
+
+export interface PortTypes {
+    [id: string]: NodeType;
+}
```

Comment:
I think this is incorrect. Each key in PortTypes is a function that has to be called and returns the PortBuilderType from above. I fixed my type error with:
```suggestion
  interface PortConfig {
    name?: string;
    label?: string;
    hidePort?: boolean;
    color?: string;
    controls?: any;
    noControls?: boolean;
  }

 export interface PortTypes {
    [id: string]: (config?: PortConfig) => PortBuilderType;
  }
```

The PortConfig represents the config the function can take based on what can be part of the config here https://github.com/chrisjpatty/flume/blob/b7e6f7a495fbd8424a827afebd9e9cf538643c13/src/typeBuilders.js#L99-L105